### ISSUE TITLE
Systematic error of ~0.4 hartree in CASCI calculations using fix_spin_

### ISSUE THREAD

**Issue by davidsousarj** [pyscf/pyscf#1965]
When I perform CASCI calculations for excited states using the `fix_spin_` method, I always get a systematic error of about 0.4 hartree in all energies. Also, this only happens with quintet (S=2) states, I also tried S = 1 and S = 0, but those are fine.

I will show two examples. First, the benzene molecule:

```
from pyscf import gto, scf, mcscf
from pyscf.tools import mo_mapping

mol = gto.M(atom='''\
C   -1.2131   -0.6884    0.0000
C   -1.2028    0.7064    0.0000
C   -0.0103   -1.3948    0.0000
C    0.0104    1.3948    0.0000
C    1.2028   -0.7063    0.0000
C    1.2131    0.6884    0.0000
H   -2.1577   -1.2244    0.0000
H   -2.1393    1.2564    0.0000
H   -0.0184   -2.4809    0.0000
H    0.0184    2.4808    0.0000
H    2.1394   -1.2563    0.0000
H    2.1577    1.2245    0.0000''', basis='ccpvdz')

mf = scf.RHF(mol)
mf.kernel()
pz_pop = mo_mapping.mo_comps('C 2pz', mol, mf.mo_coeff)
cas_list = pz_pop.argsort()[-6:]

mycas = mcscf.CASCI(mf, 6, (5,1)) # quintet state
mo = mycas.sort_mo(cas_list, base=0)
mycas.fcisolver.nroots = 10
mycas.fcisolver.spin = 4
mycas.kernel(mo)```

For this input I have got this:
```
CASCI converged
CASCI state   0  E = -230.449789575849  E(CI) = -6.10698904551344  S^2 = 6.0000000
CASCI state   1  E = -230.362043913085  E(CI) = -6.01924338274961  S^2 = 6.0000000
CASCI state   2  E = -230.362019506396  E(CI) = -6.01921897606030  S^2 = 6.0000000
CASCI state   3  E = -230.220021905087  E(CI) = -5.87722137475160  S^2 = 6.0000000
CASCI state   4  E = -230.220010150093  E(CI) = -5.87720961975782  S^2 = 6.0000000
CASCI state   5  E = -230.184717849278  E(CI) = -5.84191731894282  S^2 = 6.0000000
CASCI state   6  E = -230.184699899023  E(CI) = -5.84189936868762  S^2 = 6.0000000
CASCI state   7  E = -230.115346921931  E(CI) = -5.77254639159611  S^2 = 12.0000000
CASCI state   8  E = -230.107452498904  E(CI) = -5.76465196856893  S^2 = 6.0000000
CASCI state   9  E = -230.079031990318  E(CI) = -5.73623145998235  S^2 = 6.0000000
```
With one spin-contaminated state. If I add the line `mycas.fix_spin_(ss=4)` before the last one, then I get this:

```
CASCI converged
CASCI state   0  E = -230.049789575847  E(CI) = -5.70698904551350  S^2 = 6.0000000
CASCI state   1  E = -229.962043913083  E(CI) = -5.61924338274966  S^2 = 6.0000000
CASCI state   2  E = -229.962019506394  E(CI) = -5.61921897606035  S^2 = 6.0000000
CASCI state   3  E = -229.820021905085  E(CI) = -5.47722137475162  S^2 = 6.0000000
CASCI state   4  E = -229.820010150091  E(CI) = -5.47720961975779  S^2 = 6.0000000
CASCI state   5  E = -229.784717849277  E(CI) = -5.44191731894287  S^2 = 6.0000000
CASCI state   6  E = -229.784699899021  E(CI) = -5.44189936868767  S^2 = 6.0000000
CASCI state   7  E = -229.707452498903  E(CI) = -5.36465196856898  S^2 = 6.0000000
CASCI state   8  E = -229.679031990316  E(CI) = -5.33623145998237  S^2 = 6.0000000
CASCI state   9  E = -229.619880137340  E(CI) = -5.27707960700593  S^2 = 6.0000000
```
Observe that all CI roots are approximately 0.4 hartree higher in energy than the previous result.

Second  example, the FeS molecule:
```
from pyscf import gto, scf, mcscf
from pyscf.mcscf import avas

mol = gto.M(atom='Fe 0 0 0; S 0 0 2.2', basis='def2svp')

mf = scf.RKS(mol)
mf.xc = 'b3lyp'
mf.init_guess = 'vsap'
mf = scf.newton(mf)
mf.kernel()

ncas, nelecas, mo = avas.avas(mf, ['Fe 3d', 'Fe 4s', 'S 3p'])

mycas = mcscf.CASCI(mf, 9, (8,4))
mycas.fcisolver.spin = 4
mycas.fcisolver.nroots = 10
mycas.kernel(mo)
```
Output:
```
CASCI converged
CASCI state   0  E = -1659.60516133215  E(CI) = -33.8497464760744  S^2 = 6.0000002
CASCI state   1  E = -1659.60145486696  E(CI) = -33.8460400108820  S^2 = 6.0000000
CASCI state   2  E = -1659.60085206017  E(CI) = -33.8454372040926  S^2 = 6.0000000
CASCI state   3  E = -1659.58776168083  E(CI) = -33.8323468247509  S^2 = 6.0000000
CASCI state   4  E = -1659.58320808504  E(CI) = -33.8277932289698  S^2 = 11.9999972
CASCI state   5  E = -1659.58014959787  E(CI) = -33.8247347417946  S^2 = 6.0000000
CASCI state   6  E = -1659.57970207742  E(CI) = -33.8242872213496  S^2 = 6.0000006
CASCI state   7  E = -1659.57899951599  E(CI) = -33.8235846599157  S^2 = 6.0000025
CASCI state   8  E = -1659.57775096572  E(CI) = -33.8223361096409  S^2 = 6.0000322
CASCI state   9  E = -1659.57742159117  E(CI) = -33.8220067350985  S^2 = 6.0000308
```
Output including the line `mycas.fix_spin_(ss=4)` before the last line:
```
CASCI state   0  E = -1659.14425888025  E(CI) = -33.3934550353911  S^2 = 6.0027054
CASCI state   1  E = -1659.14270690341  E(CI) = -33.3919030585500  S^2 = 6.0004231
CASCI state   2  E = -1659.13336736205  E(CI) = -33.3825635171859  S^2 = 6.0015445
CASCI state   3  E = -1659.12177097241  E(CI) = -33.3709671275476  S^2 = 6.0023713
CASCI state   4  E = -1659.11612669673  E(CI) = -33.3653228518651  S^2 = 6.0016389
CASCI state   5  E = -1659.11582721048  E(CI) = -33.3650233656233  S^2 = 6.0016158
CASCI state   6  E = -1659.10106133918  E(CI) = -33.3502574943163  S^2 = 6.0004013
CASCI state   7  E = -1659.09970039070  E(CI) = -33.3488965458391  S^2 = 6.0004424
CASCI state   8  E = -1659.07274171709  E(CI) = -33.3219378722335  S^2 = 6.0358458
CASCI state   9  E = -1659.07003304579  E(CI) = -33.3192292009251  S^2 = 6.0388188
```
For this one there is a systematic error of 0.4-0.5 hartree for each state.

It seems that `fix_spin_` is not a reliable method to avoid spin contamination. My solution for now is using DMRG (Block2) as `FCISolver`, which has spin-adapted algorithm and therefore avoids spin contamination.

**Comment by shivupa** [comment]
This is not unreasonable since in a basis of determinants to fix the spin one usually uses a spin purification method. 

From the [documentation](https://pyscf.org/user/mcscf.html):
> `fix_spin_()` is an energy-penalty method. It may affect the efficiency and accuracy of the CASCI and CASSCF algorithms. It should not be used unless the MCSCF program has difficulties in finding the wanted spin state.

Without reading the code, PySCF must use a spin purification method like [this](https://pubs.acs.org/doi/10.1021/acs.jpca.2c01338) or [this](https://pubs.acs.org/doi/abs/10.1021/acs.jctc.7b00466)

PySCF experts feel free to weigh in if this is incorrect

**Comment by shivupa** [comment]
Yup sorry wasn't trying to re-explain the method, I misunderstood the issue! In that case shouldn't the argument to `fix_spin_` be the target $S^2$ not the target $S$

**Comment by davidsousarj** [comment]
@shivupa now that I look with attention, yes, the argument in `fix_spin_` should be $S^2$. I have redone the calculations with `mycas.fix_spin_(ss=6)` and now the results seem correct.

I thought that it followed the same convention to define `mol.spin`, which requires $2S$ as argument. This is not clear in the documentation. Also, I did not even suspect that this is the cause of the error because the calculations converge to the right spin states, $S = 2$ and $S^2 = 6$.

**Comment by davidsousarj** [comment]
I suggest to the developers that if the user input a `ss` value in `fix_spin` that did not correspond to a integer or half-integer value of `S`, at least a warning should be shown, so the user do not confuse $S^2$ and $2S$ in the input.

For example, if the input is `mycas.fix_spin_(ss=4)`, as was mine, it outputs:

"Warning: S**2 = 4 means S = 1.5616, inconsistent spin value" or something like that.

**Comment by davidsousarj** [comment]
A even user-friendlier suggestion would be allowing the user to input $S$ or $S^2$, as they wish. Then the function internally converts to the desired value.
For example, input `mycas.fix_spin_(s=2)` and the function internally converts `ss=s*(s+1)`, or 
`mycas.fix_spin_(ss=6)`, in that case `s=(sqrt(4*ss+1)-1)/2`.

**Comment by verena-neufeld** [comment]
Maybe this could be a `feature request`?
